### DISCUSSION TITLE
How to return a custom type with methods from an async function?

### DISCUSSION THREAD

**Discussion by joseluisq** [mlua-rs/mlua#454] [Q&A] [answered]
Hi, I'm playing around with `mlua` and can not find an example of returning a custom type (a struct with methods).

```rs
pub struct LuaClientResponse(pub u16);

impl UserData for LuaClientResponse {
    fn add_methods<'lua, M: UserDataMethods<'lua, Self>>(methods: &mut M) {
        methods.add_method("status", |_, resp, ()| Ok(resp.0));
    }
}

let fetch_json = lua.create_async_function(|_, uri: String| async move {
    let resp = http_client::get(&uri).await.into_lua_err()?;
    let json = resp.json::<serde_json::Value>().await.into_lua_err()?;
    let status = resp.url().to_string();
    // ...

    // This below does not work if I implement `UserData`.
    // But it does if, for example, I serialize the json payload
    // via `LuaSerdeExt` and return it like `lua.to_value(&json)`.
    Ok(LuaClientResponse(status))
})
lua.globals().set("fetch_json", fetch_json)?;
```

I would like to be able to do something like this:

```lua
local resp = fetch_json("http://localhost/xyz")
print(resp.status())
```

Is there a way to return a type with methods similar to what `UserData` does when passed as an argument in a function call?

**Comment by khvzak** [comment]
> Is there a way to return a type with methods similar to what UserData does when passed as an argument in a function call?

You can return `AnyUserData` and then take the underlying Rust object from it, see [`AnyUserData::take`](https://docs.rs/mlua/latest/mlua/struct.AnyUserData.html#method.take).
Another option - return [`UserDataRef<T>`](https://docs.rs/mlua/latest/mlua/struct.UserDataRef.html) which will act as a reference to your object.

**Comment by joseluisq** [comment, marked as answer]
It worked by calling `into_lua` on the custom data type before returning it like

```rust
let fetch = lua.create_async_function(|_, uri: String| async move {
    let resp = http_client::get(&uri).await.into_lua_err()?;
    // ...
    let status = resp.status().as_u16();
    Ok(LuaClientResponse(status).into_lua(lua))
})
```

And calling the method like this

```lua
local resp = fetch("http://localhost/xyz")
print(resp:status())
-- 200
```

Thanks!
